### ISSUE TITLE
Difference in means estimators for design with some pairs and some larger blocks

### ISSUE THREAD

**Issue by lukesonnet** [DeclareDesign/estimatr#158]
Macartan suggested exploring using the pooling estimator for matched pairs variance for a case with some pairs and some blocks.

**Comment by lukesonnet** [comment]
From @macartan:

I am seeing that formula 3.16 on p 77 of Green/Gerber. It is easy to construct examples where it is way off; eg

library(DeclareDesign)
pop <- declare_population(
  N = 40,
  block = rep(1:20, each = 2),
  Y0 = 0,
  Y1 = block,
  Z  = block_ra(block),
  Y = Z*Y1 + (1-Z)*Y0
)
# Here the standard ​deviation of the s​ampling distribution of the estimated ​ATE is 0; ​we will estimate
10.5 every time
difference_in_means(Y ~ Z, data = pop())
But the variance of the estimated
​block level ​ates will be large

with(pop(), var((Y1-Y0)[Z==1]))


Suggestion 1, for the mixed case:
If indeed formula 3.16 has something going for it then it CAN be used even if blocks are of mixed size. There is no reason for it to blow up. 

So better than :
Error in difference_in_means(Y ~ Z, blocks = block, data = pop2()) : 
  `blocks` must either all have two units/`clusters` (i.e., a matched pairs design) or all have more than two units. You cannot mix blocks of size two with blocks of a larger size. In order to estimate treatment effects with this design, ​​use inverse propensity score weights with lm_robust()

Might be: Warning: some blocks have only two units and so the Neyman formula cannot be applied. Instead se's are calculated based on variance in block level estimated ates.

or something like that. In any case 3.16 ​​can be applied.

​​Aside: I am not sure I would direct people necessarily to IPW since they may not have weights
​; also this seems to suggest getting rid of the block information.
​use block fixed effects and inverse propensity score weights with lm_robust()
Or perhaps say: you can remove the blocks argument, coarsen blocks, or use an IPW approach. 

​​Aside: lm_ln also blows up with matched pairs but without a good warning message:
lm_lin(Y ~ Z, ~ as.factor(block), data = pop())
Warning message:
In add_cis_pvals(return_frame, alpha, ci && se_type != "none") :
  Some degrees of freedom have been estimated as negative or zero
p-values and confidence intervals may not be calculated

Suggestion 2, for the mixed case:
I was also wondering if one might want to "coarsen up" ; eg take all the blocks that are just pairs and pool them to make a large block; note that if each of these has one unit in treatment and one in control then the coarsened block will also have half in treatment and half in control. With these coarsened blocks one can then apply Neyman everywhere.
One could imagine finer coarsenings in which pairs o
​f
pairs are pooled. I don't know the conventional wisdom on this however
​ so this, while possibly more efficient might be riskier than Suggestion 1.​

eg here you do much better coarsening blocks than you do by ignoring  them
​ (though you would do even better with lm_robust and ​block fixed effects)
; I wonder ​how it would do if you used
​Eq ​3.16 here

**Comment by nfultz** [comment]
Usually we would assume the average effect of Z doesn't change across blocks / depend on anything else - in this case you included the interaction you correctly recover the parameters - 

```
> lm(Y~Z:block, pop())

Call:
lm(formula = Y ~ Z:block, data = pop())

Coefficients:
(Intercept)      Z:block  
          0            1  
```

Within your sample, I think 10.5 is the correct wrong number.

**Comment by macartan** [comment]
Hi Neal -- not sure I am following. We generally want to allow
heterogeneous treatment effects. In this example the treatment effects are
1,2,3.... , 20, which averages to 1.5; the interest is the ATE not the
interaction. More generally this is just an example of a case with
heterogeneity between pairs and not within them which means that the
sampling distribution of the treatment effect has no variance but our ses
will not know that. Nothing new in that point I think.

**Comment by nfultz** [comment]
I think it was just an ugly example, nm. But is ate even useful when an
interaction is in the dgp?

**Comment by macartan** [comment]
Hi Neal -- thanks. Yes, we generally need to assume that the dgp is
unknown; then dd lets one assess how something performs given a postulated
dgp. In this case the interaction is not relevant, perhaps useful to think
of the pair level effects as being draws from some distributio; the key
thing only is that the potential outcomes are the same within a pair.

The bigger q for us though is that flawed as it is it is still a common
approach, so should we default to it also in the case where we have a
mixture of blocks of size 2 and greater.

**Comment by lukesonnet** [comment]
Okay, I think the solution for the next CRAN push is to implement Macartan's first suggestion. If there are any blocks of size 2, always use this estimator, and warn if there are also blocks that are larger than size 2 that we are doing this.

Will implement by end of day unless someone says "NO!"

**Comment by lukesonnet** [comment]
Closed by #184.
